src/search/landmarks/exploration.cc

```diff
@@ -104,115 +105,117 @@ void Exploration::build_unary_operators(const OperatorProxy &op) {
+        }
+
+        // case (b)
+        EffectsProxy effects = (op.op_or_axiom_id < 0)
+            ? task_proxy.get_axioms()[-op.op_or_axiom_id].get_effects()
+            : task_proxy.get_operators()[op.op_or_axiom_id].get_effects();
+        for (EffectProxy effect : effects) {
+            if (effect.get_fact().get_pair() == fact
+                && effect.get_conditions().empty()) {
+                return true;
+            }
+        }
     }
+    return false;
 }
 
 void Exploration::relaxed_exploration() {
     while (!prop_queue.empty()) {
-        pair<int, ExProposition *> top_pair = prop_queue.pop();
-        int distance = top_pair.first;
-        ExProposition *prop = top_pair.second;
+        Proposition *prop = prop_queue.front();
+        prop_queue.pop_front();
 
-        int prop_cost = prop->h_max_cost;
-        assert(prop_cost <= distance);
-        if (prop_cost < distance)
-            continue;
-        const vector<ExUnaryOperator *> &triggered_operators = prop->precondition_of;
+        const vector<UnaryOperator *> &triggered_operators = prop->precondition_of;
         for (size_t i = 0; i < triggered_operators.size(); ++i) {
-            ExUnaryOperator *unary_op = triggered_operators[i];
-            if (unary_op->excluded) // operator is not applied
+            UnaryOperator *unary_op = triggered_operators[i];
+            if (unary_op->excluded)
                 continue;
             --unary_op->unsatisfied_preconditions;
-            unary_op->h_max_cost = max(prop_cost + unary_op->base_cost,
-                                       unary_op->h_max_cost);
             assert(unary_op->unsatisfied_preconditions >= 0);
             if (unary_op->unsatisfied_preconditions == 0) {
-                enqueue_if_necessary(unary_op->effect, unary_op->h_max_cost);
+                enqueue_if_necessary(unary_op->effect);
             }
         }
     }
 }
 
-void Exploration::enqueue_if_necessary(ExProposition *prop, int cost) {
-    assert(cost >= 0);
-    if (prop->h_max_cost == -1 || prop->h_max_cost > cost) {
-        prop->h_max_cost = cost;
-        prop_queue.push(cost, prop);
+void Exploration::enqueue_if_necessary(Proposition *prop) {
+    if (!prop->reached) {
+        prop->reached = true;
+        prop_queue.push_back(prop);
     }
 }
 
-void Exploration::compute_reachability_with_excludes(
-    vector<vector<int>> &lvl_var,
+vector<vector<bool>> Exploration::compute_relaxed_reachability(
     const vector<FactPair> &excluded_props,
     const unordered_set<int> &excluded_op_ids) {
-    // Perform exploration using h_max-values.
-    setup_exploration_queue(task_proxy.get_initial_state(), excluded_props, excluded_op_ids);
+    setup_exploration_queue(task_proxy.get_initial_state(),
+                            excluded_props, excluded_op_ids);
     relaxed_exploration();
 
-    // Copy reachability information into lvl_var.
+    // Bundle reachability information into the return data structure.
```

Comment:
One follow-up issue Clemens and me discussed is having a proper class for this reachability information. This class could then potentially take on some responsibilities that are currently in the rhw factory and util, such as `possibly_reaches_lm` from util

Reply:
Does this work with the plan of making landmarks abstract and polymorphic? That would normally imply that logic related to things like "does something reach a landmark?" needs to live with the landmarks so that different kinds of landmarks can do this differently.